### PR TITLE
WIP: Add window opacity slider tool

### PR DESCRIPTION
**Pull request by personalizedrefrigerator**
## Summary
_see also_ #3312.

 * Implements suggestion 10 from #3259
Screenshot of semi-transparent Xournal++

## To-do
 * [x] Error handling for when the user doesn't have a compositor running.
 * [x] Remove code duplication (maybe `ZoomSlider` and the opacity slider should share a slider base-class?)
     * `AbstractSliderItem` (currently used by just the opacity slider)
 * [ ] Setting for whether to make the toolbar transparent, in addition to the Xournal widget.
     * Potential reference: https://stackoverflow.com/questions/16832581/how-do-i-make-a-gtkwindow-background-transparent-on-linux

### CODE REVIEW

File: src/enums/generated/ActionGroup.generated.cpp [hunk 1]
```diff
@@ -116,7 +116,7 @@ auto ActionGroup_fromString(const string& value) -> ActionGroup {
         return GROUP_ZOOM_FIT;
     }
 
-    g_warning("Invalid enum value for ActionGroup: \"%s\"", value.c_str());
+    g_warning("Invalid enum value for ActionGroup: «%s»", value.c_str());
```
Comment (personalizedrefrigerator) [inline src/enums/generated/ActionGroup.generated.cpp:119]: This change was made by `generateConvert.php`.

File: ui/iconsColor-dark/hicolor/scalable/actions/xopp-window-opacity-slider.svg [hunk 2]
```diff
@@ -0,0 +1 @@
...normal;shape-padding:0;shape-margin:0;inline-size:0;clip-rule:nonzero;display:inline;overflow:visible;visibility:visible;isolation:auto;mix-blend-mode:normal;color-interpolation:sRGB;color-interpolation-filters:linearRGB;solid-color:#000;solid-opacity:1;vector-effect:none;fill:none;fill-opacity:1;fill-rule:nonzero;stroke:#555761;stroke-linecap:round;stroke-linejoin:miter;stroke-miterlimit:4;stroke-dasharray:none;stroke-dashoffset:0;stroke-opacity:1;color-rendering:auto;image-rendering:auto;shape-rendering:auto;text-rendering:auto;enable-background:accumulate;stop-color:#000;stop-opacity:1;opacity:1;stroke-width:.26458333" d="M3.44 293.693a.397.397 0 0 0-.397.397.397.397 0 0 0 .396.396h1.698a.397.397 0 0 0 .396-.396.397.397 0 0 0-.396-.397H3.439z"/><circle style="opacity:1;fill:#fafafa;stroke:none;stroke-width:.264583;stroke-linecap:round;stroke-miterlimit:4;stroke-dasharray:none;paint-order:stroke markers fill;stop-color:#000" cx="2.91" cy="294.09" r="1.191"/><circle style="opacity:1;fill:none;stroke:#555761;stroke-width:.264583;stroke-linecap:round;stroke-miterlimit:4;stroke-dasharray:none;paint-order:stroke markers fill;stop-color:#000" cx="2.91" cy="294.09" r="1.191"/></g></svg>
```
Comment (personalizedrefrigerator) [inline ui/iconsColor-dark/hicolor/scalable/actions/xopp-window-opacity-slider.svg:1]: For now this is just a duplicate of the zoom slider's icon (but I think it works for this).

File: src/gui/toolbarMenubar/ToolWindowOpacitySlider.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    GtkWidget* slider_ = nullptr;
+    GtkToolItem* toolItem_ = nullptr;
+    Control* control_ = nullptr;
+};
+
+// TODO: Memory management. Who is responsible for freeing slider_, toolItem_, ...?
+ToolWindowOpacitySlider::~ToolWindowOpacitySlider() = default;
+
+ToolWindowOpacitySlider::ToolWindowOpacitySlider(ActionHandler* handler, std::string id, ActionType type,
+                                                 Control* control, IconNameHelper iconNameHelper):
```
Comment (personalizedrefrigerator) [inline src/gui/toolbarMenubar/ToolWindowOpacitySlider.cpp:38]: In `ToolMenuHandler::initToolItems`, `control->getWindow()` returns `nullptr`. As such, we can't pass the `MainWindow*` to the `ToolWindowOpacitySlider` directly. I don't like giving it the entire `control` object, though...

File: src/gui/toolbarMenubar/ToolWindowOpacitySlider.cpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+auto ToolWindowOpacitySlider::Impl::getNewToolIcon() const -> GtkWidget* {
+    return gtk_image_new_from_icon_name(iconNameHelper_.iconName("window-opacity-slider").c_str(),
+                                        GTK_ICON_SIZE_SMALL_TOOLBAR);
+}
+
+void ToolWindowOpacitySlider::Impl::setEnabled(bool enabled) {
+    if (slider_) {
+        gtk_widget_set_sensitive(GTK_WIDGET(slider_), enabled);
```
Comment (personalizedrefrigerator) [inline src/gui/toolbarMenubar/ToolWindowOpacitySlider.cpp:120]: What should we do if we can't set the window's opacity ([`gtk_widget_is_composited`](https://developer.gnome.org/gtk3/stable/GtkWidget.html#gtk-widget-is-composited) returns `false` for the main window)?

One option could be to `gtk_widget_set_sensitive(GTK_WIDGET(slider_), false)` and display a warning when the user hovers over the slider.

**Comment by rolandlo** [comment]
This looks fun, but wouldn't it make more sense to just make the `XournalWidget` transparent?

File: src/enums/generated/ActionGroup.generated.cpp (hunk 1 above)
Comment (rolandlo) [reply src/enums/generated/ActionGroup.generated.cpp:119]: It looks like the `generateConvert.php` script should be updated to produce the desired result, see #2407.

File: src/gui/toolbarMenubar/ToolWindowOpacitySlider.cpp (hunk 4 above)
Comment (rolandlo) [reply src/gui/toolbarMenubar/ToolWindowOpacitySlider.cpp:127]: That sounds like a reasonable solution.

**Comment by personalizedrefrigerator** [comment]
> make the `XournalWidget` transparent?

From what I've found online, it looks easier to make the entire window transparent (So I'll have to look into how to do that...). 

For my use case, though, making the entire window transparent is (slightly) more useful, as I plan to use this feature to take notes on video recordings of lectures. Making the entire window transparent lets me see content that would otherwise be hidden by toolbars. Of course, using the `empty toolbar` is another option!

Additionally, the slider does not allow the user to make the window entirely transparent and the window becomes fully opaque again on startup. As such, this shouldn't force the user to search for an invisible slider to make the window visible again.

**Comment by kneardhead** [comment]
Man I didn't think this will happen so quick! Thanks I truely appreciate! 

The opacity thing looks very cool not gonna lie.

File: src/enums/generated/ActionGroup.generated.cpp (hunk 1 above)
Comment (personalizedrefrigerator) [reply src/enums/generated/ActionGroup.generated.cpp:119]: See #3271

**Comment by rolandlo** [comment]
> > make the `XournalWidget` transparent?
> 
> From what I've found online, it looks easier to make the entire window transparent (So I'll have to look into how to do that...).
> 
> For my use case, though, making the entire window transparent is (slightly) more useful, as I plan to use this feature to take notes on video recordings of lectures. Making the entire window transparent lets me see content that would otherwise be hidden by toolbars. Of course, using the `empty toolbar` is another option!
> 
I have a different use case in mind: displaying some image or other information in a window below Xournal++ and using it to copy the image/info to Xournal++ by drawing the outlines etc. that shine trough. In German there is the word "abpausen" for it, see [this video](https://www.youtube.com/watch?v=BZdfXpScsuU) if my explanation was not clear. While doing that, I would like to have the toolbar normally visible, because otherwise it's a bit hard to find the right tools/colors etc.

Maybe it could be made a setting, whether the whole window is made transparent or only the `XournalWidget`.

> Additionally, the slider does not allow the user to make the window entirely transparent and the window becomes fully opaque again on startup. As such, this shouldn't force the user to search for an invisible slider to make the window visible again.

**Comment by LittleHuba** [comment]
> maybe ZoomSlider and the opacity slider should share a slider base-class?

If there is duplicated code, most definitely!

File: src/gui/toolbarMenubar/ToolZoomSlider.cpp [hunk 5]
```diff
@@ -1,182 +1,158 @@
...
 
-ToolZoomSlider::ToolZoomSlider(ActionHandler* handler, std::string id, ActionType type, ZoomControl* zoom,
+constexpr double SCALE_LOG_OFFSET = 0.20753;
+
+constexpr double FINE_STEP_COUNT = 100.0;
+constexpr double COARSE_STEP_COUNT = 10.0;
+
+// TODO(personalizedrefrigerator): Update the range to reflect the min and max zoom settings.
+constexpr auto SLIDER_RANGE =
+        AbstractSliderItem::SliderRange{DEFAULT_ZOOM_MIN, DEFAULT_ZOOM_MAX, FINE_STEP_COUNT, COARSE_STEP_COUNT};
```
Comment (personalizedrefrigerator) [inline src/gui/toolbarMenubar/ToolZoomSlider.cpp:18]: This was the case before this PR, too.

**Comment by personalizedrefrigerator** [comment]
I'm going to wait for https://github.com/xournalpp/xournalpp/pull/3312 to be merged/finished before finishing up this PR.

**Pull request closed without merging**